content/advent-2015/glow-map-reduce-for-golang.md

```diff
@@ -0,0 +1,485 @@
+
+Then for each task, the driver would contact the assigned agents,
+and send a binary clone of itself, but run in executor mode. The
+driver will also tell the executor the input data locations.
+
+Started by the agents, the executor will pull the data via network
+channels, process it, and write output to local agent.
+
+The driver can send and receive data to the executors via network
+channels also.
+
+## Code structure
+For many simple cases, the word count example is enough. However,
+in order to run the code as either driver or executor, this structure
+is recommended.
+
+```go
+package main
+
+import (
+  "flag"
+
+  _ "github.com/chrislusf/glow/driver"
+  "github.com/chrislusf/glow/flow"
+)
+
+
+var (
+  f1 = flow.New()
+  f2 = flow.New()
+
+  // input or output channel if any
+  inputChan = make(chan InputType)
+  outputChan = make(chan OutputType)
+)
+
+func init() {
+  // flow definitions
+  f1.Channel(inputChan).Map(...).Reduce(...).AddOutout(outputChan)
+  f2.Slice(...).Map(...).Reduce(...)
+}
+
+func main() {
+  // these 2 lines are needed
+  // to differentiate executor mode and driver mode.
+  // Just always add them.
+  flag.Parse()
+  flow.Ready()
+
+  // start the flow
+  go f1.Run()
+
+  // feed into input channel if any
+  ...
+
+  // wait for the output channel if any
+  ...
+
+  // possibly start other flow
+  f2.Run()
+}
+
+```
+
+So basically 2 things to follow:
+
+1. the ```flag.Parse()``` and ```flow.Ready()``` need to
+be called after the main() starts.
+- the flow definitions should be inside ```init()```
+
+
+## Why define flows in init()?
+It is worth noting that I declared the flow definition in init()
+function.
+
+This is because Go currently lacks the capability to dynamically
+compile and load a piece of code. This impacts the design of Glow.
+In distributed mode, Glow needs to send the compiled binary to Glow
+Agents, and then run the binary in executor mode by adding flow id
+and task id to existing command line options.
+
+To achieve this correctly, the flows should be statically
+deterministic. The flows and the flow steps should not change given
+the same command line parameters.
+
+Go's init() is an ideal place to cleanly define flows, in one or
+multiple files.
+
+## How to make flows dynamically?
+As mentioned above, the flows are static. How to dynamically change
+the flow?
+
+Actually we do not change flows. We can just define multiple flows,
+and dynamically invoke a flow based on the results coming out of the
+flow via channels.
+
+The flow definitions can be thought as machines. Your code can have
+many machines, defined in several files' init() functions.
+Depending on your situation, you can start a machine, feed it via
+Go's channels, and read the output also via Go's channels.
+
+A typical example would be running Linear Regression until the error
+is small enough. The error can be sent back to the driver, and the
+driver can decide whether need to run one more round of regression.
```

Comment:
I think you need to change:
"the driver can decide whether need to run" -->
"the driver can decide whether it needs to run"